Don't do nonlocal in regular Jacobian thread

lindsayad opened the pull request:
Based on coverage and module testing, we don't need this.
Additionally it actually makes automatic scaling break for real inputs
because it leads to querying global dof indices for variables that don't
have global character, e.g. 'w' in phase field

This fixes the automatic scaling error reported in #20484

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/978561/) on 4e27282 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20502/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/978569/) on 4e27282 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>643a67</th>
      <th colspan=4>#20502 4e2728</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.41%</td>
      <td>82.41%</td>
      <td>+0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73391</td>
      <td>73389</td>
      <td>-2</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15669</td>
      <td>15663</td>
      <td>-6</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20502/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20502/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20502/coverage/xfem)

</details>



This comment will be updated on new commits.

SudiptaBiswas commented on framework/src/loops/ComputeJacobianThread.C:
```diff
@@ -64,14 +62,6 @@ ComputeJacobianThread::computeJacobian()
-          std::shared_ptr<NonlocalKernel> nonlocal_kernel =
-              std::dynamic_pointer_cast<NonlocalKernel>(kernel);
-          if (nonlocal_kernel)
-            kernel->computeNonlocalJacobian();
```
We don't need this? How would the nonlocal jacobians be computed then?

lindsayad replied:
In all existing uses of nonlocal kernels, the (preconditioning) matrix is something other than diagonal-only. `ComputeJacobianThread` is only used for diagonal-only matrices. `ComputeFullJacobianThread` is used in all other cases, so for instance if you add a `SingleMatrixPreconditioner` then you will get a `ComputeFullJacobianThread`, which still calls `kernel->computeNonlocalJacobian()`

SudiptaBiswas replied:
I see. So, maybe we should at least add a warning here, so that in the future if someone tries to use nonlocal kernels with a diagonal-only preconditioner is notified about this.

lindsayad replied:
I can add an error

lindsayad committed:
Don't do nonlocal in regular Jacobian thread

Coverage testing indicates that we never hit these lines that I'm removing.
Additionally they actually makes automatic scaling break for real inputs
because it leads to querying global dof indices for variables that don't
have global character, e.g. 'w' in phase field. This is due to the fact
that `ComputeJacobianForScalingThread` inherits from and uses `ComputeJacobianThread`
`onElement` and `onBoundary`

This fixes the automatic scaling error reported in #20484

moosebuild commented:
Job [End user mac](https://civet.inl.gov/job/983605/) on 0bea252 : invalidated by @lindsayad

lindsayad commented:
Ok @dschwen, @SudiptaBiswas I think this is ready

SudiptaBiswas commented on test/tests/userobjects/shape_element_user_object/tests:
```diff
@@ -79,4 +79,27 @@
+      recover = false
+      max_threads = 1
+    []
+  []
```
This looks good to me. Should we add another test with `automatic_scaling = true`, making sure that we don't hit the error by accident?

lindsayad replied:
Oh yea ... I think that we will indeed hit that error 😆 I believe I can find a good work-around for that

dschwen reviewed (approved):
Looks good to me
